### DISCUSSION TITLE
WS2812 timer control

### DISCUSSION THREAD

**Discussion by pas940** [arendst/Tasmota#19950] [Q&A]
Hi, 
So I have a esp32 with a ws2812 RGB led on gpio 2. It works with the sliders.
I want to set up rules then use the timer to trigger them. I want to usse it as a trafic light to tell the kids when it's time to wake up / stay in bed. 
I'm not sure how to do the rules, or if this is the way to do this with out extrenal control.
Any help with correct rules to use would be great!
Thanks
Paul

**Comment by barbudor** [comment]
You can use the timer to trigger a rule
Each timer is associated with a dedicated trigger  `ON Clock#Timer=n DO something ENDON`  where `n` is the timer index 1..10
So for example to set the led as RED when triggered by Timer 5 and GREEN when triggered by Timer 6 use
```
Rule1 ON Clock#Timer=5 DO Color #FF0000 ENDON ON Clock#Timer=6 DO Color #00FF00 ENDON
Rule1 1
```
You can have almost as many `ON Clock#Timer+n DO something ENDON` as needed in a single Rule

A trick to make it easy to change colors assocated to a Timer: create once for all hte 10 triggers for each timer and use `%mem1%`...`%mem10%` variables:
```
Rule1 ON Clock#Timer=1 DO Color %mem1% ENDON ON Clock#Timer=2 DO Color %mem2% ENDON ... ON Clock#Timer=10 DO Color %mem10% ENDON
Rule1 1
```
Then you can change the color for each timer by just changing associated `%memx%`:
```
mem1 #FF0000
mem2 #00FF00
...
```

**Reply by barbudor** [reply]
Alternatively, if you need more than 10 timer, on ESP32 you have the Berry language which include a CRONTAB feature = a scheduler with very flexible syntax that can support much more than 10 events per day
Berry language is close to the Python programming language

**Comment by pas940** [comment]
Thank you. this works, but... I want it dimmer
I tried:
`Rule1 ON Clock#Timer=1 DO Color #FF0000 Dimmer 50 ENDON ON Clock#Timer=2 DO Color #00FF00 Dimmer 50 ENDON`
This doesn't work at all...
so I tried:

`Rule1 ON Clock#Timer=1 DO Color #220000 ENDON ON Clock#Timer=2 DO Color #002200 ENDON`
this works for red but not green!, red is dim just visable, and green apears to be full brightness. Any ideas what i'm doing wrong now.
![image](https://github.com/arendst/Tasmota/assets/14799544/c97c8585-dda3-44e4-afb6-f7b15efc26c9)
![image](https://github.com/arendst/Tasmota/assets/14799544/878a2555-f878-4b1b-b196-7508f81751b5)

Thanks
Paul

**Reply by pas940** [reply]
I split it into 2 rules and it works now.

**Reply by barbudor** [reply]
If you want to send 2 commands in the same DO, you need to use Backlog like
```
Rule1 ON Clock#Timer=1 DO Backlog Color #FF0000 ; Dimmer 50 ENDON
```

**Reply by pas940** [reply]
Thanks, Still having issues with the brightness not being what i expect. Seems to be rembering old rules, so as I adjust the level it uses the old one. Tried Rule1 0 Rule2 0. Do these delete the rule or just switch it off? I'll try the code above tonight. Thanks for your help

**Reply by barbudor** [reply]
`Rule1 0` disable it but doesn't erase it
To erase use `Rule1 "`

**Reply by pas940** [reply]
getting closer...

The rule you gave works but the led goes full bright for a flash (wake kids up maybe?) before being set to the dimmed value. I assume this is the time between comands been processed so I tried in the oposite order which gives rather unusal behaviour:
`19:15:00.967 RUL: CLOCK#TIMER=1 performs "Backlog Dimmer 50 ; Color #FF0000"
19:15:00.986 RSL: RESULT = {"POWER":"ON","Dimmer":50,"Color":"008000","HSBColor":"120,100,50","Channel":[0,50,0]}
19:15:01.189 RSL: RESULT = {"POWER":"ON","Dimmer":100,"Color":"FF0000","HSBColor":"0,100,100","Channel":[100,0,0]}
19:16:00.903 RUL: CLOCK#TIMER=2 performs "Backlog Dimmer 50 ; Color #00FF00"
19:16:00.920 RSL: RESULT = {"POWER":"ON","Dimmer":50,"Color":"800000","HSBColor":"0,100,50","Channel":[50,0,0]}
19:16:01.121 RSL: RESULT = {"POWER":"ON","Dimmer":100,"Color":"00FF00","HSBColor":"120,100,100","Channel":[0,100,0]}`

 Full brightness! I guess I'll have to try with the flash of full brightness

**Reply by sfromis** [reply]
To reduce risk of flashes, you can make processing much faster using `Backlog0` to avoid the default `Backlog` inter-command delay of 200 ms (SetOption34), or use `SetOption20` to allow "Update of Dimmer/Color/CT without turning power on", and then explicitly turn on when ready.

Note that mixing usage of command `Dimmer` with specific `Color` values overriding the dimming is a somewhat contradictory practice.

**Reply by pas940** [reply]
Perfect! thanks for your help guys
